Define polyfilled T_* constants from Tokenizer as int

fredden opened the pull request:
# Description

The value of these constants are not stable, and therefore already cannot be relied upon. This is because the specific values that PHP assigns can change with different versions of PHP. PHPCS does not use the values of these constants (other than to look up their name using the Tokens::tokenName() method).

There are other tools which also poly-fill these constants. Some of those tools also perform validation on the value for these constants. In order to play nicely with the arbitrary validation that other tools perform on these constants, we are switching from string values to integer values.

The [PHP manual] suggests "using big numbers like 10000" for poly-filled T_* constants. We have arbitrarily chosen to start our numbering scheme from 135_000.

All PHPCS 'native' tokens currently have reliable values. In line with PHP T_* constants, the values of these tokens should never be relied upon. In a future version of PHPCS, the values for these tokens will switch from strings to integers.

Existing tests already cover the use of these constants and do not require adjustment for the code being changed here.

[PHP manual]: https://www.php.net/manual/en/tokens.php

## Suggested changelog entry
- Change: poly-filled PHP T_* constants now have integer values. This makes no difference to PHPCS, but is done to increase compatibility with other tools which also poly-fill (and validate the values of) these same T_* tokens.
- Deprecate: the value of PHPCS native T_* constants should no longer be regarded as stable. In a future version their values will change from strings to integers. The Tokens::tokenName() method will continue to work when this change is made.

## Related issues/external references

Fixes #1286 

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix _(non-breaking change which fixes an issue)_
- [x] New feature _(non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that would cause existing functionality to change)_
    - [ ] This change is only breaking for integrators, not for external standards or end-users.
- [ ] Documentation improvement


## PR checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have checked there is no other PR open for the same change.
- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/HEAD/.github/CONTRIBUTING.md).
- [x] I grant the project the right to include and distribute the code under the BSD-3-Clause license (and I have the right to grant these rights).
- [x] I have added tests to cover my changes.
- [x] I have verified that the code complies with the projects coding standards.
- [ ] \[Required for new sniffs\] I have added XML documentation for the sniff.
- [ ] I have opened a sister-PR in the [documentation repository](https://github.com/PHPCSStandards/PHP_CodeSniffer-documentation) to update the Wiki.

<!--
============================================================================================
Please make sure your pull request passes all continuous integration checks!

PRs which are failing their CI checks will likely be ignored by the maintainers.

Small PRs using atomic, descriptive commits are hugely appreciated as it will make
reviewing your changes easier for the maintainers.
============================================================================================
-->

fredden committed:
Define polyfilled constants from Tokenizer as int

The value of these constants are not stable, and therefore already cannot be
relied upon. This is because the specific values that PHP assigns can change
with different versions of PHP. PHPCS does not use the values of these
constants (other than to look up their name using the Tokens::tokenName()
method).

There are other tools which also polyfill these constants. Some of those tools
also perform validation on the value for these constants. In order to play
nicely with the arbitrary validation that other tools perform on these
constants, we are switching from string values to integer values.

All PHPCS 'native' tokens currently have reliable values. In line with PHP T_*
constants, the values of these tokens should never be relied upon. In a future
version of PHPCS, the values for these tokens will switch from strings to
integers.

Existing tests already cover the use of these constants and do not require
adjustment for the code being changed here.

fredden commented:
@jrfnl the 'validate' step for "Find use of Tokens properties" check is failing on this code. Would you like me to rewrite the check or the code to avoid this?

jrfnl commented:
> @jrfnl the 'validate' step for "Find use of Tokens properties" check is failing on this code. Would you like me to rewrite the check or the code to avoid this?

What gets flagged is this:
```
src\Util\Tokens.php:135:Tokens::$polyfillMappingTable = $polyfillMappingTable;
```

I think updating the check to exclude this particular use would be most appropriate, though I'm not sure how straight-forward that would be.

fredden commented:
I have worked out a way to use a static method after all. This means we don't need to change the CI/CD check for external use of deprecated properties.

jrfnl commented on src/Util/Tokens.php:
```diff
@@ -985,4 +922,65 @@ public static function getHighestWeightedToken(array $tokens)
+        // any class constants or properties before this method has fully run,
+        // PHP will intitialise the class, leading to warnings about undefined
+        // T_* constants.
+        $tokensToPolyfill = [
```
I can see that these are ordered alphabetically.

To me as a maintainer, it makes more sense to order them by the PHP version which introduced the token (with comments between the sets annotating the PHP version).

With the current ordering, this would mean that every time there is a PHP version drop, someone would need to go through the list to figure out in which PHP version each token was introduced, while if the list is ordered by PHP version, handling a version drop is more straight forward.

fredden replied:
I have updated the order as suggested.

jrfnl commented on src/Util/Tokens.php:
```diff
@@ -606,6 +536,13 @@ final class Tokens
+    /**
+     * Mapping table for polyfilled constants
+     *
+     * @var array<int, string>
```
```suggestion
     * @var array<int, string>
```

No need to change anything, but just pointing out that we can't actually be sure/guarantee that the key will always be an `int` as - just like PHPCS did - other tooling could have polyfilled the tokens with some other type of value.

jrfnl reviewed (commented):
Hi @fredden, thanks for that update and sorry for my slow response.

Creative solution and I appreciate the inline comments about the timing aspect of the static method and the fragility of the setup. That should hopefully prevent breakage when this method is touched in the future.

The fragility is a little concerning, but if it works for normal PHPCS CLI use and the PHPCS native test setup (which it does), I'm okay with it _(though I wonder which user of an external dependency will complain first about something breaking because of this change...)_


Conceptually, I have two questions:
* The assignment to `$polyfillMappingTable[constant($tokenName)]` does not contain any protection against overwriting an existing array index.
    While token constants _should be_ unique, if another external tool does this incorrectly and that code would run before the PHPCS polyfill code, it will cause us problems: https://3v4l.org/OCE9X#veol (see the output for PHP 7.2 - 8.0)
    We may need to throw an Exception and just block the run if this happens. What do you think ?

* Should the method include protection against being called twice ? (as it is a `public static` method defining crucial information) ?
    I've not been able to come up with a scenario in which this becomes problematic, but it still feels risky.
    Also see https://3v4l.org/bdlTc - when called the second time, each polyfilled token would overwrite its previously created own entry in the `Tokens::$polyfillMappingTable`.


Nitpicky things:
* `polyfillTokenizerConstants()` - there is the PHP `Tokenizer` and the PHPCS `Tokenizer`, purely based on the method name, this could confuse people.
    `polyfillPHPNativeTokenizerConstants()` may be a bit wordy, but might be clearer ?
    You might also want to update the method doc block ? (add "PHP native" to the summary, remove "PHP native" from the `@internal` note)

fredden committed:
Reorder PHP constants for better maintainability

fredden committed:
Better protect against collisions

fredden commented:
> I wonder which user of an external dependency will complain first about something breaking because of this change.

It is because an external project decided to "validate" the value of these polyfilled constants (and complain about our historic choice of values) that we are changing the values we assign.
The values will be internally consistent within any particular invocation of PHP, but there's no guarantee that they will have the same value next time around.
PHP itself has different values for some of these constants across versions.

> The assignment to `$polyfillMappingTable[constant($tokenName)]` does not contain any protection against overwriting an existing array index.
> While token constants _should be_ unique, if another external tool does this incorrectly and that code would run before the PHPCS polyfill code, it will cause us problems: https://3v4l.org/OCE9X#veol (see the output for PHP 7.2 - 8.0)
> We may need to throw an Exception and just block the run if this happens. What do you think ?

Collisions are a problem, yes.

I have spent some time today and I think I've solved the concerns raised. I made our collision detection more robust and added a check for any collisions before we ran our definitions. These two steps should mean that any other libraries which define collisions will be detected and reported before our code starts getting confused.

> Should the method include protection against being called twice ? (as it is a `public static` method defining crucial information) ?
> I've not been able to come up with a scenario in which this becomes problematic, but it still feels risky.
> Also see https://3v4l.org/bdlTc - when called the second time, each polyfilled token would overwrite its previously created own entry in the `Tokens::$polyfillMappingTable`.

No. Calling the method more than once should not cause any problems. I don't see value in adding protection against this.

> `polyfillTokenizerConstants()` - there is the PHP `Tokenizer` and the PHPCS `Tokenizer`, purely based on the method name, this could confuse people.
> `polyfillPHPNativeTokenizerConstants()` may be a bit wordy, but might be clearer ?
> You might also want to update the method doc block ? (add "PHP native" to the summary, remove "PHP native" from the `@internal` note)

This name was specifically chosen to allow for us to move the PHP_CodeSniffer tokeniser constants into this method in future.

jrfnl commented on src/Util/Tokens.php:
```diff
@@ -991,4 +923,99 @@ public static function getHighestWeightedToken(array $tokens)
+            'T_PRIVATE_SET',
+            'T_PROTECTED_SET',
+            'T_PUBLIC_SET',
+            'T_READONLY',
```
Duplicate ? Introduced in 8.1 I believe.

fredden replied:
Thanks; fixed.

jrfnl reviewed (commented):
We discussed this PR in a call today and what's missing - but also very difficult to add - is tests.

We've looked into adding `phpt` based tests for this and @fredden will continue working on that.

The principle of this PR is validated and agreed upon though, so this will go into the PHPCS 4.1.0 release.

fredden committed:
Add tests for new tokenizer constant code

jrfnl commented on tests/EndToEndPhpt/Util/Tokens/polyfillTokenizerConstants-collision-php.phpt:
```diff
@@ -0,0 +1,18 @@
+--SKIPIF--
+<?php
+if (version_compare(PHP_VERSION, "8.4", ">=")) {
+    echo "skip because tokens used in this test already exist in PHP 8.4 so we cannot test polyfilling them", PHP_EOL;
```
```suggestion
if (version_compare(PHP_VERSION, '8.4', '>=') === true) {
    echo 'skip because tokens used in this test already exist in PHP 8.4 so we cannot test polyfilling them', PHP_EOL;
```

jrfnl commented on tests/EndToEndPhpt/Util/Tokens/polyfillTokenizerConstants-collision-php.phpt:
```diff
@@ -0,0 +1,18 @@
+--FILE--
+<?php
+define('T_PUBLIC_SET', T_STRING);
+require('src/Util/Tokens.php');
```
```suggestion
require 'src/Util/Tokens.php';
```

jrfnl commented on tests/EndToEndPhpt/Util/Tokens/polyfillTokenizerConstants-collision-user.phpt:
```diff
@@ -0,0 +1,20 @@
+--SKIPIF--
+<?php
+if (version_compare(PHP_VERSION, "8.4", ">=")) {
+    echo "skip because tokens used in this test already exist in PHP 8.4 so we cannot test polyfilling them", PHP_EOL;
```
```suggestion
if (version_compare(PHP_VERSION, '8.4', '>=') === true) {
    echo 'skip because tokens used in this test already exist in PHP 8.4 so we cannot test polyfilling them', PHP_EOL;
```

jrfnl commented on tests/EndToEndPhpt/Util/Tokens/polyfillTokenizerConstants-collision-user.phpt:
```diff
@@ -0,0 +1,20 @@
+define('T_PRIVATE_SET', 10000);
+define('T_PROTECTED_SET', 10000);
+define('T_PUBLIC_SET', 10000);
+require('src/Util/Tokens.php');
```
```suggestion
require 'src/Util/Tokens.php';
```

jrfnl commented on tests/EndToEndPhpt/Util/Tokens/polyfillTokenizerConstants-number.phpt:
```diff
@@ -0,0 +1,14 @@
+--SKIPIF--
+<?php
+if (version_compare(PHP_VERSION, "8.4", ">=")) {
+    echo "skip because tokens used in this test already exist in PHP 8.4 so we cannot test polyfilling them", PHP_EOL;
```
```suggestion
if (version_compare(PHP_VERSION, '8.4', '>=') === true) {
    echo 'skip because tokens used in this test already exist in PHP 8.4 so we cannot test polyfilling them', PHP_EOL;
```

jrfnl commented on tests/EndToEndPhpt/Util/Tokens/polyfillTokenizerConstants-number.phpt:
```diff
@@ -0,0 +1,14 @@
+--FILE--
+<?php
+define('T_PUBLIC_SET', 135000);
+require('src/Util/Tokens.php');
```
```suggestion
require 'src/Util/Tokens.php';
```

jrfnl commented on tests/EndToEndPhpt/Util/Tokens/polyfillTokenizerConstants-skip-names.phpt:
```diff
@@ -0,0 +1,17 @@
+--SKIPIF--
+<?php
+if (version_compare(PHP_VERSION, "8.4", ">=")) {
+    echo "skip because tokens used in this test already exist in PHP 8.4 so we cannot test polyfilling them", PHP_EOL;
```
```suggestion
if (version_compare(PHP_VERSION, '8.4', '>=') === true) {
    echo 'skip because tokens used in this test already exist in PHP 8.4 so we cannot test polyfilling them', PHP_EOL;
```

jrfnl commented on tests/EndToEndPhpt/Util/Tokens/polyfillTokenizerConstants-skip-names.phpt:
```diff
@@ -0,0 +1,17 @@
+define('T_', T_STRING); // Too short
+define('ONE', T_STRING); // First character is not 'T'
+define('TWO', T_STRING); // Second character is not '_'
+require('src/Util/Tokens.php');
```
```suggestion
require 'src/Util/Tokens.php';
```

jrfnl commented on tests/EndToEndPhpt/Util/Tokens/polyfillTokenizerConstants-skip-names.phpt:
```diff
@@ -0,0 +1,17 @@
+// Using T_STRING as a value because that would be a collision (and throw) if the constant name looked like a tokenizer constant.
+define('T_', T_STRING); // Too short
+define('ONE', T_STRING); // First character is not 'T'
+define('TWO', T_STRING); // Second character is not '_'
```
```suggestion
define('T_', T_STRING); // Too short.
define('ONE', T_STRING); // First character is not 'T'.
define('TWO', T_STRING); // Second character is not '_'.
```

jrfnl reviewed (commented):
Awesome work on the tests @fredden !

I've just left some nitpicks comments inline to make the test code largely comply with the coding standards. We're not automatically checking against the standards for `phpt` files currently, nor would that work out-of-the-box.

Might be worth opening an issue to consider scanning `phpt` files as a feature enhancement ?
In my opinion, it would be a low priority enhancement as the `phpt` file format isn't widely used, but, if nothing else, it would help this project to keep the `phpt` files in line with the rest of the code.

jrfnl commented on src/Util/Tokens.php:
```diff
@@ -612,6 +537,13 @@ final class Tokens
     ];
 
+    /**
+     * Mapping table for polyfilled constants
```
```suggestion
     * Mapping table for polyfilled constants.
```

jrfnl reviewed (commented):
One last question - the `get_defined_constants(true)` function call is repeated numerous times, including within loops, while the result won't change between calls.

Would it make sense to call the function once, store to a local variable and reference that throughout the function ?

jrfnl commented on phpunit.xml.dist:
This same change also needs to be made in the `phpunit-lte9.xml.dist` file.

jrfnl commented:
Oh... where's my brain... I keep thinking of more things... - might be good to add a section about these EndToEnd tests in the `CONTRIBUTING` file with links to the available documentation about the format ?

jrfnl commented on tests/EndToEndPhpt/Util/Tokens/polyfillTokenizerConstants-collision-user.phpt:
```diff
@@ -0,0 +1,20 @@
+Fatal error: Uncaught Exception: Externally polyfilled tokenizer constant value collision detected! T_PROTECTED_SET has the same value as T_PRIVATE_SET in %s:%d
+Stack trace:
+#0 %s(%d): PHP_CodeSniffer\Util\Tokens::polyfillTokenizerConstants()
+#1 Standard input code(%d): require('...')
```
```suggestion
#1 Standard input code(%d): require('%s')
```
(otherwise the test will fail on Windows)

jrfnl commented on tests/EndToEndPhpt/Util/Tokens/polyfillTokenizerConstants-collision-php.phpt:
```diff
@@ -0,0 +1,18 @@
+Fatal error: Uncaught Exception: Externally polyfilled tokenizer constant value collision detected! T_PUBLIC_SET has the same value as T_STRING in %s:%d
+Stack trace:
+#0 %s(%d): PHP_CodeSniffer\Util\Tokens::polyfillTokenizerConstants()
+#1 Standard input code(%d): require('...')
```
```suggestion
#1 Standard input code(%d): require('%s')
```
(otherwise the test will fail on Windows)

fredden committed:
Update tests/EndToEndPhpt/Util/Tokens/polyfillTokenizerConstants-collision-php.phpt

Co-authored-by: Juliette <663378+jrfnl@users.noreply.github.com>

fredden committed:
Update tests/EndToEndPhpt/Util/Tokens/polyfillTokenizerConstants-collision-php.phpt

Co-authored-by: Juliette <663378+jrfnl@users.noreply.github.com>

fredden committed:
Update src/Util/Tokens.php

Co-authored-by: Juliette <663378+jrfnl@users.noreply.github.com>

fredden committed:
Update tests/EndToEndPhpt/Util/Tokens/polyfillTokenizerConstants-skip-names.phpt

Co-authored-by: Juliette <663378+jrfnl@users.noreply.github.com>

fredden committed:
Update tests/EndToEndPhpt/Util/Tokens/polyfillTokenizerConstants-skip-names.phpt

Co-authored-by: Juliette <663378+jrfnl@users.noreply.github.com>

fredden committed:
Update tests/EndToEndPhpt/Util/Tokens/polyfillTokenizerConstants-skip-names.phpt

Co-authored-by: Juliette <663378+jrfnl@users.noreply.github.com>

fredden committed:
Update tests/EndToEndPhpt/Util/Tokens/polyfillTokenizerConstants-collision-user.phpt

Co-authored-by: Juliette <663378+jrfnl@users.noreply.github.com>

fredden committed:
Update tests/EndToEndPhpt/Util/Tokens/polyfillTokenizerConstants-collision-user.phpt

Co-authored-by: Juliette <663378+jrfnl@users.noreply.github.com>

fredden committed:
Update tests/EndToEndPhpt/Util/Tokens/polyfillTokenizerConstants-number.phpt

Co-authored-by: Juliette <663378+jrfnl@users.noreply.github.com>

fredden committed:
Update tests/EndToEndPhpt/Util/Tokens/polyfillTokenizerConstants-number.phpt

Co-authored-by: Juliette <663378+jrfnl@users.noreply.github.com>

fredden committed:
Add End2EndPhpt test suite to legacy configuration

fredden committed:
Update tests/EndToEndPhpt/Util/Tokens/polyfillTokenizerConstants-collision-php.phpt

Co-authored-by: Juliette <663378+jrfnl@users.noreply.github.com>

fredden committed:
Update tests/EndToEndPhpt/Util/Tokens/polyfillTokenizerConstants-collision-user.phpt

Co-authored-by: Juliette <663378+jrfnl@users.noreply.github.com>

fredden committed:
Avoid calling function in loops

fredden committed:
Fix test for PHP < 8.0

fredden committed:
Fix test for test suite paths

jrfnl reviewed (commented):
Status update: we discussed this PR in a call today.

The tests fails when running with Xdebug in coverage mode.
This is most likely due to PHPUnit preloading files when running in coverage mode.

Then again, it may also have something to do with the `tests/bootstrap.php` file being preloaded, though if that's the case, the tests without Xdebug should also fail.

Tests in `phpt` mode already **_always_** run in a separate process, so there is nothing we can toggle on that end.

So, considering all this, we are leaning towards moving the `phpt` test suite config to its own configuration file so it will be more straight-forward to run all the "unit/integration" tests in the normal `test` workflow, while running the phpt end to end tests separately (always).

In part, the reason for this is that PHPUnit does not have a `--exclude-testsuite` CLI option, which means that if everything would stay in one PHPUnit config file, we would need to run every test suite separately, making things more complicated when merging code coverage reports.

Action plan:
* Remove the `testsuite` config from the pre-existing PHPUnit config files and move it to a `endtoend.xml.dist` file, preferably with only `<phpunit>`  top-level attributes which are PHPUnit cross-version compatible, so we can avoid having to have separate config files for PHPUnit 8-9 vs 10 and higher.
* Add a separate PHPUnit test step to the `end-to-end-tests.yml` workflow which runs just the PHPUnit end to end tests.
    We may need to check in the "normal" PHPUnit `test.yml` workflow if any other steps/config needs to be copied into the `end-to-end-tests.yml` for this to work properly.
* Mention the new tests and how to run them in the `CONTRIBUTING` guide. This doesn't need to be a full write-up. It can be expanded later on, but it should have an initial mention in the PR which introduces these new tests.

fredden committed:
Avoid running phpt tests with coverage

When PHPUnit runs in coverage mode, it pre-loads all files. The files being
tested in the End2EndPhpt test suite need to be able to run code _before_ any
PHP_CodeSniffer classes are loaded, this isn't compatible. To avoid this, we
are moving this test suite to the 'end to end' test stage, where we do not run
code coverage.

fredden committed:
Introduce phpt tests in contributing guide

jrfnl commented on .github/workflows/end-to-end-tests.yml:
```diff
@@ -48,6 +48,17 @@ jobs:
+      - name: Install Composer dependencies
+        uses: "ramsey/composer-install@3cf229dc2919194e9e36783941438d17239e8520" # 3.1.1
+        with:
+          composer-options: ${{ matrix.php == '8.6' && '--ignore-platform-req=php+' || '' }}
```
```suggestion
          composer-options: ${{ matrix.php == 'nightly' && '--ignore-platform-req=php+' || '' }}
```

fredden replied:
Good catch. This was a copy/paste error on my part. Fixed in f11637eb06bceb795c5297bc3caf03c0deb48b0c

jrfnl commented on .github/workflows/end-to-end-tests.yml:
```diff
@@ -48,6 +48,17 @@ jobs:
+          custom-cache-suffix: $(date -u "+%Y-%m")
+
+      - name: 'PHPUnit: run end-to-end tests (phpt)'
+        run: php "vendor/bin/phpunit" -c phpunit-e2e.xml.dist --no-coverage
```
Re: `--no-coverage` - For the record, I don't think this is needed, though it also doesn't do any harm.

For PHPUnit to run in code- coverage mode two conditions need to be fullfilled:
1. Xdebug needs to be available (which it is not in this workflow due to `coverage: none` for setup-php)
2. In the PHPUnit config file, a `<filter>` (PHPUnit < 9.3) or `<coverage>` (PHPUnit 9.3+) configuration needs to be setup (which the `phpunit-e2e.xml.dist` file does not include).

fredden replied:
I have added a comment to explain why we have this flag in 595cae8fe8b65bccaa39799f883c8603c3f14828

jrfnl commented on phpunit-e2e.xml.dist:
```diff
@@ -0,0 +1,18 @@
+    convertErrorsToExceptions="true"
+    convertWarningsToExceptions="true"
+    convertNoticesToExceptions="true"
+    convertDeprecationsToExceptions="true"
```
Darn... these attributes are not PHPUnit cross-version compatible (not available in PHPUnit 10 and higher).
Also, for similar behaviour in PHPUnit 10 and higher, we need different attributes.

How do we want to handle this ?
* Have two separate config files (PHPUnit 8/9 and PHPUnit 10/11) like for the regular tests ?
* Remove these attributes from the config and set these type of options via the command-line in the GHA workflow ?

fredden replied:
Removed in c39cb5b324c82962cf6cfddeb09b0fa533bfd7d7. These are all covered by unit tests.

jrfnl replied:
> These are all covered by unit tests.

To clarify for anyone wondering:
* We **_do_** want CI to fail if the code which is being changed in this PR would end up generating PHP errors/warnings/notices or deprecations.
* PHPUnit changed the attributes which "rule" what PHP notices will fail tests in PHPUnit 10, making the highlighted settings not PHPUnit cross-version compatible.
* PHPUnit defaults to **not** failing on PHP deprecation notices since a particular PHPUnit 9.x version. This behaviour is perpetuated and enhanced in PHPUnit 10, where PHP deprecation notices will now be hidden by default.
* This is dangerous for PHPCS as PHPCS is often run in the CI of other projects and may fail _their_ CI if PHPCS fails due to deprecation notices.
* In other words, we need a way to safeguard that if a change in PHP would result in a deprecation notice for the code changed in this PR, it will cause _our_ CI to fail.

Now given that:
* the code changed in this PR is**_not_** covered via unit tests...
* but it is _run_ in the test bootstrap of the unit tests (also run via PHPUnit)
* and that the unit tests already has different configs for the various PHPUnit versions to ensure that all PHP deprecation notices are shown and will fail builds....

... we have verified that PHPUnit will exit with a non-zero exit code if a PHP deprecation notice is encountered in the bootstrapping of our unit tests - which it does.

And as it does, we don't need to worry about the E2E phpt tests not failing on such a deprecation notice and we can therefore use a simpler, PHPUnit cross-version compatible config.

jrfnl commented on .github/workflows/end-to-end-tests.yml:
I think, for now, having both these tests (bash + phpt) in one job will be fine, though we may want to split this into two jobs at a later point in time, as with the current setup the phpt tests failing could hide another failure in the bash tests.

Having said that, if this is one job... the job name on line 19 feels wrong.

(Oh and I also noticed an out of place / old copy/paste artifact on line 32 which can be removed - yamllint annotation)

fredden replied:
Split in f806de49898f2de1dc9aca526f2d7cdb881e78a3. Stray comment removed in f11637eb06bceb795c5297bc3caf03c0deb48b0c

jrfnl reviewed (changes requested):
Thanks for the update @fredden !

Gone through the latest changes and left a few more remarks, but we're nearly there.

fredden committed:
Use 'nightly' version to match matrix

fredden committed:
Add code comment to explain use of '--no-coverage'

fredden committed:
Remove settings which are not cross-version compatible

While it would be great to have these here, the simplicity of supporting all
PHPUnit versions in one file is more valuable to us at this stage.

We have checked that PHP deprecations in the bootstrap stage are reported when
running PHPUnit unit tests.

fredden committed:
Split GitHub Action workflow into separate jobs

fredden committed:
Merge remote-tracking branch 'upstream/4.x' into issue-1286/polyfill-tokenizer-const-as-int

fredden committed:
Refine test names

jrfnl reviewed (approved):
Awesome! Thanks @fredden !

I will merge this once the 4.0.2 release has gone out. 

Note to self: **squash**merge this PR + update (and review) branch protection rules when I merge this.
